Modules/_zstd/_zstdmodule.c

```diff
@@ -0,0 +1,968 @@
+        }
+    }
+
+    /* Unknown parameter */
+    if (name == NULL) {
+        PyOS_snprintf(pos_msg, sizeof(pos_msg),
+                      "unknown parameter (key %d)", key_v);
+        name = pos_msg;
+    }
+
+    /* Get parameter bounds */
+    if (is_compress) {
+        bounds = ZSTD_cParam_getBounds(key_v);
+    } else {
+        bounds = ZSTD_dParam_getBounds(key_v);
+    }
+    if (ZSTD_isError(bounds.error)) {
+        PyErr_Format(state->ZstdError,
+                     "Zstd %s parameter \"%s\" is invalid. (zstd v%s)",
+                     type, name, ZSTD_versionString());
+        return;
+    }
+
+    /* Error message */
+    PyErr_Format(state->ZstdError,
+                 "Error when setting zstd %s parameter \"%s\", it "
+                 "should %d <= value <= %d, provided value is %d. "
+                 "(zstd v%s, %d-bit build)",
+                 type, name,
+                 bounds.lowerBound, bounds.upperBound, value_v,
+                 ZSTD_versionString(), 8*(int)sizeof(Py_ssize_t));
+}
+
+/* -------------------------
+     Train dictionary code
+   ------------------------- */
+
+/*[clinic input]
+_zstd._train_dict
+
+        samples_bytes: PyBytesObject
+            Concatenation of samples.
+        samples_size_list: object(subclass_of='&PyList_Type')
+            List of samples' sizes.
+        dict_size: Py_ssize_t
+            The size of the dictionary.
+        /
+
+Internal function, train a zstd dictionary on sample data.
+[clinic start generated code]*/
+
+static PyObject *
+_zstd__train_dict_impl(PyObject *module, PyBytesObject *samples_bytes,
+                       PyObject *samples_size_list, Py_ssize_t dict_size)
+/*[clinic end generated code: output=ee53c34c8f77886b input=fdb00ebe474d1df9]*/
+{
+    Py_ssize_t chunks_number;
+    size_t *chunk_sizes = NULL;
+    PyObject *dst_dict_bytes = NULL;
+    size_t zstd_ret;
+    Py_ssize_t sizes_sum;
+    Py_ssize_t i;
+
+    /* Check arguments */
+    if (dict_size <= 0) {
+        PyErr_SetString(PyExc_ValueError, "dict_size argument should be positive number.");
+        return NULL;
+    }
+
+    chunks_number = Py_SIZE(samples_size_list);
+    if ((size_t) chunks_number > UINT32_MAX) {
+        PyErr_SetString(PyExc_ValueError,
+                        "The number of samples should <= UINT32_MAX.");
+        return NULL;
+    }
+
+    /* Prepare chunk_sizes */
+    chunk_sizes = PyMem_Malloc(chunks_number * sizeof(size_t));
+    if (chunk_sizes == NULL) {
+        PyErr_NoMemory();
+        goto error;
+    }
+
+    sizes_sum = 0;
+    for (i = 0; i < chunks_number; i++) {
+        PyObject *size = PyList_GET_ITEM(samples_size_list, i);
+        chunk_sizes[i] = PyLong_AsSize_t(size);
+        if (chunk_sizes[i] == (size_t)-1 && PyErr_Occurred()) {
+            PyErr_SetString(PyExc_ValueError,
+                            "Items in samples_size_list should be an int "
+                            "object, with a size_t value.");
```

Comment:
If this exception can be triggered by Python code, we should avoid using C types in it.